compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -452,6 +452,11 @@ class TreeChecker extends Phase with SymTransformer {
       tree1
     }
 
+    override def typedWhileDo(tree: untpd.WhileDo)(implicit ctx: Context): Tree = {
+      assert((tree.cond ne EmptyTree) || ctx.phase.refChecked, i"invalid empty condition in while at $tree")
+      super.typedWhileDo(tree)
+    }
```

Comment:
And here is the addition of an explicit `Ycheck` test that `WhileDo(EmptyTree, _)` is not used too early, as requested by @smarter.